packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -1469,7 +1478,7 @@ function performSyncWorkOnRoot(root: FiberRoot) {
   flushPassiveEffects();
 
   let lanes = getNextLanes(root, NoLanes);
-  if (!includesSyncLane(lanes)) {
+  if (!includesSyncLaneOrForceSync(lanes, root)) {
```

Comment:
Instead of using performSyncWorkOnRoot for this, I wonder if you can just use performConcurrentWorkOnRoot without yielding. That way you don't have to mess with the lanes and other special cases for sync.

We already have cases for performing concurrent work without yielding.

https://github.com/facebook/react/blob/a48e54f2b74d87d90b0a3bff3a9b52fd9188308b/packages/react-reconciler/src/ReactFiberWorkLoop.js#L1047-L1052